Adding autocompletion for commands and while installing packages

timsschoen opened the pull request:
I added some code and a small bash script to allow for bash autocompletion while typing commands and while installing packages.

I think this makes working with ughub a little more comfortable, especially if you are accustomed to using the linux shell.

stephanmg commented:
Hey @timsschoen,

neat idea. Can we do the same for Sh (and possibly this can be done for Zsh too)?

By the way: I would suggest to the user in the `README.md` to add it to the user local `~/.bash_completion` completion file (As suggested by the Bash FAQ here: https://github.com/scop/bash-completion/blob/master/README.md) instead of the system-wide in `/etc/`. (Also users might not have access on other computers and also you enable the completion for *ALL* users on your computer. While this is not a problem for your personal computer, basically single-user, it will be a problem for all multi-user systems to force everybody to use the completion. For some hints and pitfalls see about enabling the completion see [this](https://serverfault.com/questions/506612/standard-place-for-user-defined-bash-completion-d-scripts)).

Another comment: This ties `ughub` with Bash, so I would make sure, that it does not break on Windows w/o Bash. As to my understanding `ughub` does not require Bash but the standard `CMD.exe` only. (Otherwise less platform-independence)

Best,
Stephan

timsschoen commented:
Hi @stephanmg,

thank you for the feedback.

I changed to installation notice for bash users to use ~/.bash_completion.

Using ughub on windows should not be changed at all, as windows users use the bat file to start the python script.

Integrating in zsh looks like it should work with [bashcompinit](https://stackoverflow.com/a/8492043) as it currently is.
I'm not sure adding this to sh is easily possible.
Additionally I added support for autocompletions using the fish shell.

The changes are in [this commit](https://github.com/timsschoen/ughub/commit/42321abdc401bd08b123aa0866ee70d5fe9db0f1).

stephanmg commented:
Ah @timsschoen yes, sorry for the Windows comment. You are right in this case.

Well, do you have experience with auto-completion in Cmd.exe? It seems it is not really useful or working last time I tried to use it, so maybe we do not need this for Windows or you have other experience, let me know. Would be useful to add it for Windows as well.  As to my knowledge a lot of people recently use WSL to compile ug4 on Windows, so this would not really an acute reason to bother adding completion to Windows.

timsschoen commented:
No, no experience of autocompletion in cmd. But it looks like it is [not possible](https://stackoverflow.com/a/6465841).

Instead I have added support for autocomplete when using powershell with [this commit](https://github.com/timsschoen/ughub/commit/25005d8d4ce59ff3a57bddaea298c945b76cc4ef).

stephanmg commented:
> No, no experience of autocompletion in cmd. But it looks like it is [not possible](https://stackoverflow.com/a/6465841).
> 
> Instead I have added support for autocomplete when using powershell with [this commit](https://github.com/timsschoen/ughub/commit/25005d8d4ce59ff3a57bddaea298c945b76cc4ef).

That is great. Powershell is certainly more interesting right now (At least for me).

sreiter commented:
Hey, thank's a lot for such a nice contibution! And also very cool that you provided solutions for different platforms!

However, I see a problem in how the information on commands and their parameters is duplicated several times (GetAutoCompletions, fish, and powershell completion scripts). All the information is already specified in `ughubHelpContents.content` in `ughubHelpContents.py`.

I think it should be possible to simply parse the parameters for a given command from that table. This could be done in `GetAutoCompletions`, for example.

If you need a short description, you could add an entry `shortDescription` to each command in the `ughubHelpContents.content` dictionary, e.g., the one from your fish autocompletion script.

It would be great if we could use the information from `ughubHelpContents.content` for the different autocompletion scripts, e.g., for **fish**. Maybe we could add an option to `help` which prints a list of commands (`help --commands`) and another one which prints a short description only (`help someCommand --shortdescription`). The fish script could then invoke something like (**pseudo-code**):
```
set -l ughub_commands = ughub help --commands
for (c in $ughub_commands )
  complete -f -c ughub -n "not __fish_seen_subcommand_from $ughub_commands" -a $c -d ughub help $c --shortdescription
```
And similar for the powershell script.

If we would go with the current duplication of command names and arguments, it would get very tedious and prone to error if one would add a command or rename a parameter in ughub.

One last thing: Since you added a new command (`get-completions`), you should add some documentation on it to `ughubHelpContents.content` in `ughubHelpContents.py`.

It would be great if you could evaluate that approach. Maybe it is not much work after what you already achieved.

Thanks again for taking the time to contribute!

timsschoen commented:
Thank you for the feedback and nice ideas.

I now added a command "ughub help --commands", which returns all available commands, and "ughub help CMD --shortdescription" with a short description, which will be automatically shown in fish autocomplete.

No solution has any hardcoded autocompletes now, everything is parsed out of ughubHelp.

I replaced the extra command "ughub get-packagenames" by a new argument "--namesonly", which can be passed to "ughub list", so that only the packagenames are returned.

sreiter commented on ughub.py:
```diff
@@ -1030,7 +1079,10 @@ def RunUGHub(args):
 			print("All rights reserved")
-
+		
+		elif cmd == "get-completions":
```
All the other commands are just concatenated words without separators. So maybe `getcompletions` would be more consistent with the other commands.

sreiter commented on ughubHelpContents.py:
```diff
@@ -328,21 +353,32 @@
 		},
 
+		{
+			"name": "get-completions",
```
As noted above: `getcompletions` would be more consistent with the other commands.

sreiter commented on ughubHelp.py:
```diff
@@ -69,33 +84,85 @@ def PrintCommands():
-def PrintCommandHelp(cmdName):
+def PrintCommandHelp(cmdName, args=[]):
+
+	shortdesc = ughubUtil.HasCommandlineOption(args, ("--shortdescription",))
```
Thinking about it... maybe `--short` would suffice? I think `ughub help list --short` is probably expressive enough. This only concerns the command line argument, not the internal variables and fields - here `shortdescription` / `shortdesc` is better, I think.

sreiter commented on ughubHelpContents.py:
```diff
@@ -100,7 +103,18 @@
+					"description":	"Prints all available commands, separated by a space."
+				},
+				{
+					"name": "--shortdescription",
```
As noted above, this option could probably be simply called `--short`.

sreiter reviewed (approved):
Thanks again for this cool and solid contribution. I took the liberty add some very small comments - no real issues. Maybe you could have a quick look at them.
- I think replacing `get-completions` with `getcompletions` would probably make sense considering a consistent command name format.
- Please decide yourself whether `--short` would be a better option name to `help` than `--shortdescription`. This only concerns the command line option name, not the internal variables/fields.

Thanks!

timsschoen commented:
Sure thing, I changed the command names and fixed a small error, which I seem to have added in one of the last commits.
